client/app/components/composites/MenuPriority/MenuPriority.js

```diff
@@ -0,0 +1,166 @@
+import { Component, PropTypes } from 'react';
+import ReactDOM from 'react-dom';
+import r, { div, a } from 'r-dom';
+import _ from 'lodash';
+import classNames from 'classnames';
+import { canUseDOM } from '../../../utils/featureDetection';
+
+import Menu from '../Menu/Menu';
+import css from './MenuPriority.css';
+
+const EXTRA_SPACING_RIGHT = 24;
+const LINK_SPACING = 18;
+const ROUNDING_ERROR_MARGIN = 2;
+
+class MenuPriority extends Component {
+
+  constructor(props, context) {
+    super(props, context);
+
+    this.links = _.sortBy(this.props.content, ['priority']);
+    this.componentDidMount = this.componentDidMount.bind(this);
+    this.updateWidths = this.updateWidths.bind(this);
+    this.updateNav = this.updateNav.bind(this);
+
+    this.state = {
+      priorityWrapperWidth: '0px',
+      priorityLinks: this.links.filter((l) => l.priority >= 0),
+      hiddenLinks: this.links,
+    };
+  }
+
+  componentDidMount() {
+    if (canUseDOM) {
+      const that = this;
+      const priorityMenu = ReactDOM.findDOMNode(that);
+      let linksWithBreakPoints = null;
+
+      const menuIsVisible = function menuIsVisible(c) {
+        // This is a naive check for visibility
+        // PriorityMenu and its parent might change display settings due to responsive layout
+        return window.getComputedStyle(c, null).getPropertyValue('display') !== 'none' &&
+          window.getComputedStyle(c.parentNode, null).getPropertyValue('display') !== 'none';
+      };
+
+      // Wait for a paint to be done before calculating offsetWidths and stuff
+      // ComponentDidMount is called after React component is passed to DOM,
+      // but painting is not necessarily ready yet at that point
+      window.requestAnimationFrame(() => {
+        if (linksWithBreakPoints == null && menuIsVisible(priorityMenu)) {
+          linksWithBreakPoints = that.updateWidths(that.links);
+          that.updateNav(linksWithBreakPoints);
+        }
+      });
+
+      window.addEventListener('resize', () => {
```

Comment:
This handler will potentially create tons of layout trashes since the `updateWidths` method queries the offset dimensions for each link, forcing a rerender for each iteration.

Let's keep an eye on the performance of this, and add some throttling etc. if things become unbearable.